### PR TITLE
Marking scheme design

### PR DESCRIPTION
**Pull request by GitGeminiano**
Hello everyone,

I modified both the front end and back end to be able to choose the criteria visibility among the different users. Please have a look at it.

Regards,
Christian Garcia

### CODE REVIEW

File: app/controllers/graders_controller.rb [hunk 1]
```diff
@@ -280,7 +280,7 @@ def criteria_with_assoc(assignment, options = {})
     criterion_ids = options[:criterion_ids]
     includes = options[:includes] || CRITERION_ASSOC
 
-    criteria = assignment.get_criteria.includes(includes)
+    criteria = assignment.get_criteria.empty? ? assignment.get_criteria : assignment.get_criteria.includes(includes)
```
Comment (david-yz-liu) [inline app/controllers/graders_controller.rb]: What is the purpose of this change?

File: app/models/assignment.rb [hunk 2]
```diff
@@ -723,10 +723,22 @@ def criterion_class
   end
 
   def get_criteria
+    is_peer_review? ? get_peer_visible_criteria : get_ta_visible_criteria
```
Comment (david-yz-liu) [inline app/models/assignment.rb]: The `is_peer_review` check is not correct. Add a parameter which takes a symbol, either `:all`, `:ta`, and `:peer`, and returns the correct criteria in each of these three cases.

File: app/models/result.rb [hunk 3]
```diff
@@ -126,8 +126,9 @@ def unrelease_partial_results
   end
 
   def check_for_nil_marks
-    num_criteria = submission.assignment.rubric_criteria.count +
-                   submission.assignment.flexible_criteria.count
+    peer_review = submission.assignment.is_peer_review?
+    num_criteria = filter_criteria(submission.assignment.rubric_criteria, peer_review).count +
```
Comment (david-yz-liu) [inline app/models/result.rb]: Use the `get_criteria` method in the Assignment model rather than this "filter_criteria".

File: app/views/rubrics/_criterion_editor.html.erb [hunk 4]
```diff
@@ -38,6 +38,20 @@
      <%= t('criteria_edit_total_mark', criterion_total_mark: criterion.weight*4) %>
     </span>
     <br>
+      <h4><span><%= I18n.t(:visibility) %></span></h4>
```
Comment (david-yz-liu) [inline app/views/rubrics/_criterion_editor.html.erb]: In views, use `t` instead of `I18n.t`

File: spec/models/assignment_spec.rb [hunk 5]
```diff
@@ -136,6 +136,71 @@
...
+                                                   assignment: @assignment,
+                                                   ta_visible: false,
+                                                   peer_visible: true) }
+            @ta_and_peer_criteria = Array.new(2) { create(:rubric_criterion,
+                                                          assignment: @assignment,
+                                                          peer_visible: true) }
+            end
+
+          it 'shows the criteria visible to tas only' do
+            expect(@assignment.get_criteria.length).to eq 4
```
Comment (david-yz-liu) [inline spec/models/assignment_spec.rb]: Check the actual criteria ids

File: spec/models/assignment_spec.rb [hunk 6]
```diff
@@ -1035,7 +1100,7 @@
   describe '#grade_distribution_as_percentage' do
     before :each do
       @assignment = create(:assignment)
-      5.times { create(:rubric_criterion, assignment: @assignment) }
+      5.times { create(:rubric_criterion, assignment: @assignment, peer_visible: true) }
```
Comment (david-yz-liu) [inline spec/models/assignment_spec.rb]: Calculating Assignment grades should not require `peer_visible`criteria to be successful. Does this method `grade_distribution_as_percentage` need to be updated as well?

File: spec/models/result_spec.rb [hunk 7]
```diff
@@ -14,7 +14,7 @@
 
   describe '.student_marks_by_assignment' do
     let(:assignment) { create(:assignment) }
-    let!(:criteria) { Array.new(2) { create(:rubric_criterion, assignment: assignment) } }
+    let!(:criteria) { Array.new(2) { create(:rubric_criterion, assignment: assignment, peer_visible: true) } }
```
Comment (david-yz-liu) [inline spec/models/result_spec.rb]: Same comment as previous one.

File: app/controllers/graders_controller.rb (hunk 1 above)
Comment (GitGeminiano) [inline app/controllers/graders_controller.rb]: One of the test cases causes assignment.get_criteria to be empty. The includes method raises an error when applied to an empty array. That is why I added this check.

File: app/views/rubrics/_criterion_editor.html.erb (hunk 4 above)
Comment (GitGeminiano) [inline app/views/rubrics/_criterion_editor.html.erb]: I will change it. Thank you for noticing it.

File: test/functional/results_controller_test.rb [hunk 8]
```diff
@@ -465,7 +465,11 @@ def teardown
 
           should 'and the result is available' do
             SubmissionFile.make(:submission => @submission)
-            Mark.make(:result => @result)
+            if @assignment.marking_scheme_type == Assignment::MARKING_SCHEME_TYPE[:rubric]
+              Mark.make(:rubric, :result => @result)
```
Comment (david-yz-liu) [inline test/functional/results_controller_test.rb]: Use the new hash syntax `result: @result`

File: app/views/flexible_criteria/_criterion_editor.html.erb [hunk 9]
```diff
@@ -41,9 +41,21 @@
...
     <span><%= t('flexible_criteria.description_hint') %></span>
+    <br>
+    <br>
+    <h4><span><%= I18n.t(:visibility) %></span></h4>
+    <h5>
+      <%= f.check_box :ta_visible,
+                      onclick: 'set_onbeforeunload(true);' %>
+      <%= f.label :ta_visible,
+                  t('flexible_criteria.ta_visible') %>
+      <%= f.check_box :peer_visible,
```
Comment (david-yz-liu) [inline app/views/flexible_criteria/_criterion_editor.html.erb]: Put a line break between the two checkboxes.

File: app/models/rubric_criterion.rb [hunk 10]
```diff
@@ -2,6 +2,7 @@
 
 class RubricCriterion < Criterion
   self.table_name = 'rubric_criteria' # set table name correctly
+  include ActiveModel::Validations
```
Comment (david-yz-liu) [inline app/models/rubric_criterion.rb]: I don't think this is still necessary, is it?

**Comment by david-yz-liu** [comment]
@GitGeminiano I noticed that the **flexible criteria** "mark as complete" button bug is still there (it seems to check all criteria, rather than all ta-visible criteria.

File: test/functional/results_controller_test.rb (hunk 8 above)
Comment (GitGeminiano) [inline test/functional/results_controller_test.rb]: I changed it. Thanks. I noticed many other instances in this file in which we use the old syntax. Is this something I should change? Is it worth making an issue for?

File: app/models/rubric_criterion.rb (hunk 10 above)
Comment (GitGeminiano) [inline app/models/rubric_criterion.rb]: Not anymore. Thanks.

**Comment by david-yz-liu** [comment]
@GitGeminiano Thanks! I made a few minor changes, which you can see in the latest commit.
